### PR TITLE
Fixed a hidden quadratic runtime in hashed collect

### PR DESCRIPTION
**Pull request by goedderz**
### Scope & Purpose

AQLs hashed COLLECT variant (`COLLECT ... OPTIONS { method: "hash" }`) had a hidden quadratic runtime (in the number of results) due to an `std::distance` call on an `std::unordered_map`.

- [X] :hankey: Bugfix 
- [x] :book: CHANGELOG entry made
- [X] :muscle: The behavior in this PR was *manually tested*
- [X] :computer: The behavior change can be verified via automatic tests: https://github.com/arangodb/simple-performance-test/pull/32

#### Backports:

- [X] Backport required for **3.7** (still **TODO**)

### Testing & Verification

- [X] This change is a trivial rework / code cleanup without any test coverage.
- [X] This change is already covered by existing tests, such as *shell_server_aql* and gtest *HashedCollect.\**.

Additionally:

- [X] There are tests in an external testing repository: https://github.com/arangodb/simple-performance-test/pull/32

### Documentation

- [x] Added a *Changelog Entry* (referencing the corresponding public or internal issue number)

### CODE REVIEW

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11813/

File: arangod/Aql/HashedCollectExecutor.h [hunk 1]
```diff
@@ -205,6 +205,8 @@ class HashedCollectExecutor {
   std::vector<Aggregator::Factory> _aggregatorFactories;
 
   GroupKeyType _nextGroupValues;
+
+  size_t _returnedGroups = 0;
```
Comment (jsteemann) [inline arangod/Aql/HashedCollectExecutor.h:209]: @goedderz: Does this variable ever need to be reset to zero, or is it fine to be ever-increasing for one HashedCollectExecutor instance?

Comment (goedderz) [reply arangod/Aql/HashedCollectExecutor.h:209]: @jsteemann: That would be necessary in an `initializeCursor` method, if it existed. As it doesn't exist, the object will be destructed and constructed between subquery-iterations.

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/Performance/job/perf-simple-branch/9/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11818/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/11824/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/11839/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11906/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11909/

image

**Review by maierlars — APPROVED** [review]

**Pull request merged**